File: Programmers/258709. 주사위 고르기/HodaeSsi.java
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.IntStream;

class Solution {
    private static int[] answer = {};
    private static int max = Integer.MIN_VALUE;

    public int[] solution(int[][] dice) {

        combination(dice.length / 2, 0, dice, new java.util.ArrayList<>(), 0);

        return answer;
    }

    // 조합의 수
    private void combination(int maxDepth, int depth, int[][] diceList, List<Integer> stack, int peekIdx) {
        if (depth >= maxDepth) {
            // A, B 다이스 조합
            List<Integer> aList = stack;
            List<Integer> bList = new java.util.ArrayList<>();

            for (int i = 0; i < diceList.length; i++) {
                if (!aList.contains(i)) {
                    bList.add(i);
                }
            }

            // 각자 모든 합 계산(B는 오름차순 정렬)
            List<Integer> aSums = new java.util.ArrayList<>();
            List<Integer> bSums = new java.util.ArrayList<>();

            findAllSum(
                    IntStream.range(0, diceList.length)
                            .filter(aList::contains)
                            .mapToObj(idx -> diceList[idx])
                            .toArray(int[][]::new),
                    0, 0, aSums);
            findAllSum(
                    IntStream.range(0, diceList.length)
                            .filter(bList::contains)
                            .mapToObj(idx -> diceList[idx])
                            .toArray(int[][]::new),
                    0, 0, bSums);

            Collections.sort(bSums);

            // 이분탐색
            int aWin = 0;
            for (int i = 0; i < aSums.size(); i++) {
                int target = aSums.get(i);
                int temp = upperBound(bSums, target);
                aWin += temp;
            }

            if (aWin > max) {
                max = aWin;
                answer = new int[stack.size()];
                for (int j = 0; j < stack.size(); j++) {
                    answer[j] = stack.get(j) + 1;
                }
            }

            if (aWin > max) {
                max = aWin;
                answer = new int[stack.size()];
                for (int j = 0; j < stack.size(); j++) {
                    answer[j] = stack.get(j) + 1;
                }
            }

            return ;
        }

        for (int i = peekIdx; i < diceList.length; i++) {
            stack.add(i);
            combination(maxDepth, depth + 1, diceList, stack, i + 1);
            stack.remove(stack.size() - 1);
        }
    }

    private void findAllSum(int[][] diceList, int depth, int sum, List<Integer> results) {
        if (depth >= diceList.length) {
            results.add(sum);
            return ;
        }

        for (int i = 0; i < diceList[0].length; i++) {
            sum += diceList[depth][i];
            findAllSum(diceList, depth + 1, sum, results);
            sum -= diceList[depth][i];
        }
    }

    // target보다 작은 값의 최대 인덱스
    private int upperBound(List<Integer> list, int target) {
        int left = 0;
        int right = list.size() - 1;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (list.get(mid) >= target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }
}

File: Programmers/258707. n + 1 카드게임/HodaeSsi.py
# 풀이 방법 : 코인을 가장 적게 사용하는 방식으로 도달 가능한 최대 라운드 계산(greedy)
    # 코인에 대해 그리디하게 접근 하는 경우, 카드를 해당 라운드에서 사용하진 않지만 코인을 사용하여 킵 해 놓는 경우도 커버 됨
# 시간 복잡도 : O(n^2) (n은 카드의 개수)
def solution(coin, cards):
    answer = 0

    card_coin = {} # key - 카드 번호, value - [필요 코인 개수, 사용 여부]
    for i in range(len(cards) // 3):
        card_coin[cards[i]] = [0, False]

    for i in range(1, len(cards) // 3 + 1): # 최대 진행 가능 라운드는 cards 개수의 1/3
        find_pair = False
        card_coin[cards[len(cards) // 3 + 2 * (i - 1)]] = [1, False]
        card_coin[cards[len(cards) // 3 + 2 * (i - 1) + 1]] = [1, False]

        for j in range(len(cards) // 3 + 2 * (i - 1) + 2):
            left = card_coin.get(cards[j], [-100, False])
            right = card_coin.get(len(cards) + 1 - cards[j], [-100, False])
            if left[0] == -100 or left[1] or right[0] == -100 or right[1]:
                continue

            if left[0] == 0 + right[0] == 0:
                card_coin[cards[j]][1] = True
                card_coin[len(cards) + 1 - cards[j]][1] = True
                find_pair = True
                break

        if find_pair:
            answer = i
            continue
        if coin < 1:
            break

        for j in range(len(cards) // 3 + 2 * (i - 1) + 2):
            left = card_coin.get(cards[j], [-100, False])
            right = card_coin.get(len(cards) + 1 - cards[j], [-100, False])
            if left[0] == -100 or left[1] or right[0] == -100 or right[1]:
                continue

            if left[0] + right[0] == 1:
                card_coin[cards[j]][1] = True
                card_coin[len(cards) + 1 - cards[j]][1] = True
                find_pair = True
                coin -= 1
                break

        if find_pair:
            answer = i
            continue
        if coin < 2:
            break

        for j in range(len(cards) // 3 + 2 * (i - 1) + 2):
            left = card_coin.get(cards[j], [-100, False])
            right = card_coin.get(len(cards) + 1 - cards[j], [-100, False])
            if left[0] == -100 or left[1] or right[0] == -100 or right[1]:
                continue

            if left[0] + right[0] == 2:
                card_coin[cards[j]][1] = True
                card_coin[len(cards) + 1 - cards[j]][1] = True
                find_pair = True
                coin -= 2
                break

        if find_pair:
            answer = i
            continue
        else:
            break

    return answer + 1